src/dmd/backend/aarray.d

```diff
@@ -0,0 +1,545 @@
+     */
+
+    Value[] values()
+    {
+        if (!nodes)
+            return null;
+
+        const aligned_keysize = aligntsize(Key.sizeof);
+        auto p = cast(Value *)malloc(nodes * Value.sizeof);
+        assert(p);
+        auto q = p;
+        foreach (e; buckets)
+        {
+            while (e)
+            {
+                memcpy(q, cast(void*)(e + 1) + aligned_keysize, Value.sizeof);
+                ++q;
+                e = e.next;
+            }
+        }
+        return p[0 .. nodes];
+    }
+
+    /********************************************
+     * Rehash an array.
+     */
+
+    void rehash()
+    {
+        //printf("Rehash\n");
+        if (!nodes)
+            return;
+
+        size_t newbuckets_length = prime_list[$ - 1];
+
+        foreach (prime; prime_list[0 .. $ - 1])
+        {
+            if (nodes <= prime)
+            {
+                newbuckets_length = prime;
+                break;
+            }
+        }
+        auto newbuckets = cast(aaA**)calloc(newbuckets_length, (aaA*).sizeof);
+        assert(newbuckets);
+
+        foreach (e; buckets)
+        {
+            while (e)
+            {
+                auto en = e.next;
+                auto b = &newbuckets[e.hash % newbuckets_length];
+                e.next = *b;
+                *b = e;
+                e = en;
+            }
+        }
+
+        free(buckets.ptr);
+        buckets = null;
+        buckets = newbuckets[0 .. newbuckets_length];
+    }
+
+
+    /*********************************************
+     * For each element in the AArray,
+     * call dg(Key* pkey, Value* pvalue)
+     * If dg returns !=0, stop and return that value.
+     */
+
+    int apply(int delegate(Key*, Value*) dg)
+    {
+        if (!nodes)
+            return 0;
+
+        //printf("AArray.apply(aa = %p, keysize = %d, dg = %p)\n", &this, Key.sizeof, dg);
+
+        const aligned_keysize = aligntsize(Key.sizeof);
+
+        foreach (e; buckets)
+        {
+            while (e)
+            {
+                auto result = dg(cast(Key*)(e + 1), cast(Value*)(e + 1) + aligned_keysize);
+                if (result)
+                    return result;
+                e = e.next;
+            }
+        }
+
+        return 0;
+    }
+
+  private:
+
+    aaA*[] buckets;
+    size_t nodes;               // number of nodes
+    TKey tkey;
+}
+
+private:
+
+/**********************************
+ * Align to next pointer boundary, so value
+ * will be aligned.
+ */
+
+size_t aligntsize(size_t tsize)
+{
+    // Is pointer alignment on the x64 4 bytes or 8?
+    return (tsize + size_t.sizeof - 1) & ~(size_t.sizeof - 1);
+}
+
+immutable uint[14] prime_list =
+[
+    97U,         389U,
+    1543U,       6151U,
+    24593U,      98317U,
+    393241U,     1572869U,
+    6291469U,    25165843U,
+    100663319U,  402653189U,
+    1610612741U, 4294967291U
+];
+
+/***************************************************************/
+
+// Simple values
+public struct Tinfo(K)
```

Comment:
Is this used?

Reply:
It will be by anyone making a simple hash table with a simple key. Currently it is only used by the test code further down.